airflow/models/taskinstance.py

```diff
@@ -1495,8 +1495,7 @@ def handle_failure(
             except Exception:
                 self.log.exception('Failed to send email to: %s', task.email)
 
-        if not test_mode:
-            session.merge(self)
+        session.merge(self)
```

Comment:
Is this related?

Reply:
Yes, it can't write task_instance info into DB/session into ut in the original logic.

Reply:
But I assume there is a reason it doesn’t save the task instance, although I do not know what. What would fail if you remove this change?

Reply:
> But I assume there is a reason it doesn’t save the task instance, although I do not know what. What would fail if you remove this change?

I think so at the beginning, and I have tried to find the historical git log, nothing found. The UT always successful if remove this change although the logic is wrong.

Reply:
Find the one...
<pre>
commit aa01fab4906df3175e5e6589ec0095a5586279db
Author: Maxime Beauchemin <maxime.beauchemin@apache.org>
Date:   Mon Jan 12 21:54:40 2015 +0000

    Improving the test command line subcommand

<pre>

Reply:
I add a config `dag_file_processor.UNIT_TEST_MODE = False` in UT as a workaround. It works well but I wonder whether it is normative.